### ISSUE TITLE
Spark tutorial uses sortByKey(), which isn't implemented in PySpark

### ISSUE THREAD

**Issue by JoshRosen** [amplab/training#7]
The Spark tutorial current uses sortByKey(), but this isn't currently implemented in PySpark.

**Comment by shivaram** [comment]
Will not fix for the dry run on Feb 5th...

**Comment by andyk** [comment]
Just curious, when are we hoping to get this fixed by? Is this something I could help with?

**Comment by JoshRosen** [comment]
It shouldn't be too hard to fix if you're familiar with PySpark's internals.  The implementation will be analogous to shuffle's implementation and can probably be modeled after Scala Spark's sortByKey() method.

I've been a little busy, so I probably won't have time to work on this before the weekend.

**Comment by shivaram** [comment]
Just checking if we have a fix for this .. Or should we consider removing this top-50 exercise from the python setup ?

**Comment by JoshRosen** [comment]
Implementing sortByKey() is blocked by [porting sample() to PySpark](https://spark-project.atlassian.net/browse/SPARK-651), since the RangePartitioner relies on sampling to determine the ranges.  A naive implementation could be very inefficient, since it needs to first evaluate the RDD to determine the sample size, then sample the RDD to determine the ranges, and finally shuffle it.

From a quick inspection of the code, it looks like the Scala RangePartitioner might be inefficient because it doesn't appear to cache the RDD even though it performs two actions on it.

**Comment by shivaram** [comment]
Thanks for the update. I guess for the tutorial we can just add a note saying this exercise does not apply for python right now ?

**Comment by mateiz** [comment]
Yup, the right thing would be to pre-sort each partition and write that out to disk, then serve ranges from it (similar to the Shark PDE model). But for now if you implement it the same way in Python it won't be a problem.

**Comment by shivaram** [comment]
This is sort of a separate issue, but having takeSample() work would also be useful for k-means. Right now we just use take(K) as a substitute.

**Comment by mateiz** [comment]
I'd actually prefer to remove this part of the exercise or replace it. Can we put another question instead?

**Comment by shivaram** [comment]
Matei, Reynold and me talked about this and the plan is to filter pages with visits > 100K or something like that.
